producer-c/producer-cloudwatch-integ/CanaryStreamCallbacks.cpp

```diff
@@ -0,0 +1,216 @@
+{
+    Aws::CloudWatch::Model::PutMetricDataRequest cwRequest;
+    cwRequest.SetNamespace("KinesisVideoSDKCanary");
+    cwRequest.AddMetricData(metricDrum);
+    pCanaryStreamCallbacks->pCwClient->PutMetricDataAsync(cwRequest, OnPutMetricDataResponseReceivedHandler);
+}
+
+VOID ComputeStreamMetricsFromCanary(STREAM_HANDLE streamHandle, PCanaryStreamCallbacks pCanaryStreamCallbacks) {
+    StreamMetrics canaryStreamMetrics;
+    canaryStreamMetrics.version = STREAM_METRICS_CURRENT_VERSION;
+    getKinesisVideoStreamMetrics(streamHandle, &canaryStreamMetrics);
+    pCanaryStreamCallbacks->currentFrameRateDatum.SetValue(canaryStreamMetrics.currentFrameRate);
+    pCanaryStreamCallbacks->currentFrameRateDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::None);
+    CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->currentFrameRateDatum);
+    pCanaryStreamCallbacks->currentViewDurationDatum.SetValue(canaryStreamMetrics.currentViewDuration / HUNDREDS_OF_NANOS_IN_A_MILLISECOND);
+    pCanaryStreamCallbacks->currentViewDurationDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::Milliseconds);
+    CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->currentViewDurationDatum);
+}
+
+VOID ComputeClientMetricsFromCanary(CLIENT_HANDLE clientHandle, PCanaryStreamCallbacks pCanaryStreamCallbacks) {
+    ClientMetrics canaryClientMetrics;
+    canaryClientMetrics.version = CLIENT_METRICS_CURRENT_VERSION;
+    getKinesisVideoMetrics(clientHandle, &canaryClientMetrics);
+    pCanaryStreamCallbacks->contentStoreAvailableSizeDatum.SetValue(canaryClientMetrics.contentStoreAvailableSize);
+    pCanaryStreamCallbacks->contentStoreAvailableSizeDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::None);
+    CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->contentStoreAvailableSizeDatum);
+}
+
+VOID GetCurrentMemoryAllocation(PCanaryStreamCallbacks pCanaryStreamCallbacks) {
+    pCanaryStreamCallbacks->memoryAllocationSizeDatum.SetValue(getInstrumentedTotalAllocationSize());
+    pCanaryStreamCallbacks->memoryAllocationSizeDatum.SetUnit(Aws::CloudWatch::Model::StandardUnit::None);
+    CanaryStreamSendMetrics(pCanaryStreamCallbacks, pCanaryStreamCallbacks->memoryAllocationSizeDatum);
+}
+VOID CanaryStreamRecordFragmentEndSendTime(PStreamCallbacks pCanaryStreamCallbacks, UINT64 lastKeyFrameTime, UINT64 curKeyFrameTime) {
+    auto mapPtr = ((PCanaryStreamCallbacks) pCanaryStreamCallbacks)->timeOfNextKeyFrame;
+    (*mapPtr)[lastKeyFrameTime / HUNDREDS_OF_NANOS_IN_A_MILLISECOND] = curKeyFrameTime;
+    auto iter = mapPtr->begin();
+    while (iter != mapPtr->end()) {
+        // clean up from current timestamp of 5 min timestamps would be removed
+        if (iter->first < (GETTIME() - 300 * HUNDREDS_OF_NANOS_IN_A_SECOND) / HUNDREDS_OF_NANOS_IN_A_MILLISECOND) {
+            iter = mapPtr->erase(iter);
+        } else {
+            break;
```

Comment:
I advise folks to avoid using break and continue statements modifying the codeflow in the loops. This one is a simple case so I guess it's ok